Mathlib/NumberTheory/NumberField/CanonicalEmbedding/FundamentalCone.lean

```diff
@@ -0,0 +1,209 @@
+
+/-!
+# Fundamental Cone
+
+Let `K` be a number field of signature `(r₁, r₂)`. We define an action of the group of units
+`(𝓞 K)ˣ` of `K` on the space `ℝ^r₁ × ℂ^r₂`. The fundamental cone is a cone in `ℝ^r₁ × ℂ^r₂`
+that is a fundamental domain for the action of `(𝓞 K)ˣ` up to roots of unity.
+
+## Main definitions and results
+
+* `NumberField.mixedEmbedding.unitSMul`: The action of `(𝓞 K)ˣ` on `ℝ^r₁ × ℂ^r₂` defined, for
+`u : (𝓞 K)ˣ`, by multiplication component by component with `mixedEmbedding K u`.
+
+* `NumberField.mixedEmbedding.fundamentalCone`: A cone in `ℝ^r₁ × ℂ^r₂` --that is a subset stable
+under multiplication by a real number, see `smul_mem_of_mem`--, that is also a fundamental domain
+for the action of `(𝓞 K)ˣ` up to roots of unity, see `exists_unitSMul_me` and
+`torsion_unitSMul_mem_of_mem`.
+
+## Tags
+
+number field, canonical embedding, units
+-/
+
+variable (K : Type*) [Field K]
+
+namespace NumberField.mixedEmbedding
+
+open NumberField NumberField.InfinitePlace
+
+/-- The space `ℝ^r₁ × ℂ^r₂` with `(r₁, r₂)` the signature of `K`. -/
+local notation "E" K =>
+  ({w : InfinitePlace K // IsReal w} → ℝ) × ({w : InfinitePlace K // IsComplex w} → ℂ)
+
+noncomputable section UnitSMul
+
+/-- The action of `(𝓞 K)ˣ` on `ℝ^r₁ × ℂ^r₂` defined, for `u : (𝓞 K)ˣ`, by multiplication component
+by component with `mixedEmbedding K u`. -/
+@[simps]
+instance unitSMul : SMul (𝓞 K)ˣ (E K) where
+  smul := fun u x ↦ (mixedEmbedding K u) * x
+
+instance : MulAction (𝓞 K)ˣ (E K) where
+  one_smul := fun _ ↦ by simp_rw [HSMul.hSMul, SMul.smul, Units.coe_one, map_one, one_mul]
+  mul_smul := fun _ _ _ ↦ by simp_rw [HSMul.hSMul, SMul.smul, Units.coe_mul, map_mul, mul_assoc]
+
+instance : SMulZeroClass (𝓞 K)ˣ (E K) where
+  smul_zero := fun _ ↦ by simp_rw [HSMul.hSMul, SMul.smul, mul_zero]
+
+variable [NumberField K]
+
+theorem unitSMul_eq_iff_mul_eq {x y : (𝓞 K)} {u : (𝓞 K)ˣ} :
+    u • mixedEmbedding K x = mixedEmbedding K y ↔ u * x = y := by
+  rw [unitSMul_smul, ← map_mul, Function.Injective.eq_iff, ← Submonoid.coe_mul, Subtype.mk_eq_mk]
+  exact mixedEmbedding_injective K
+
+theorem norm_unit (u : (𝓞 K)ˣ) :
+    mixedEmbedding.norm (mixedEmbedding K u) = 1 := by
+  rw [norm_eq_norm, show |(Algebra.norm ℚ) (u : K)| = 1
+      by exact NumberField.isUnit_iff_norm.mp (Units.isUnit u), Rat.cast_one]
+
+theorem norm_unitSMul (u : (𝓞 K)ˣ) (x : E K) :
+    mixedEmbedding.norm (u • x) = mixedEmbedding.norm x := by
+  rw [unitSMul_smul, map_mul, norm_unit, one_mul]
+
+end UnitSMul
+
+noncomputable section logMap
+
+open NumberField.Units NumberField.Units.dirichletUnitTheorem FiniteDimensional
+
+variable [NumberField K] {K}
+
+/-- The map from `ℝ^r₁ × ℂ^r₂` to `{w : InfinitePlace K // w ≠ w₀} → ℝ` (with `w₀` a fixed place)
+defined in such way that: 1) it factors the map `logEmbedding`, see `logMap_eq_logEmbedding`;
+2) it is constant on the lines `{c • x | c ∈ ℝ}`, see `logMap_smul`. -/
+def logMap (x : E K) : {w : InfinitePlace K // w ≠ w₀} → ℝ := by
```

Comment:
It doesn't matter that much, but can you write this in term mode?

Reply:
Sure.